design/parallel-commit/globally-non-unique-timestamps.md

```diff
@@ -1,3 +1,27 @@
+# Allow commit_ts to be non-globally unique.
+
+This document is about the problem and the solution of the non-globally-unique `commit_ts` problem.
+
+## The problem
+
+in TiKV, rolling back a transaction need to leave a rollback record in order to guarantee consistency. But rollback records and commit records are both saved in Write CF, and the commit records are saved with `{user_key}{commit_ts}` as the internal key, while that of rollback records are `{user_key}{start_ts}`. Previously the `commit_ts`es are TSOs allocated from PD, which is guaranteed globally unique. However when we tries to use a calculated timestamp as the `commit_ts` to avoid the latency of PD's RPC, it's possible that a rollback record gets the same internal key as a commit record, but we need to keep them both.
+
+## The solution
+
+The solution is to keep the commit record, but with a `has_overlapped_rollback` flag in this case to indicate that there's a rollback happened here whose `start_ts` equals to the current record's `commit_ts`. But only "protected" rollbacks need to set the flag. Non-protected rollback records can be dropped without introducing potential inconsistency.
+
+To do this, when performing a protected rollback operation, check the records in write CF to see if there's already a commit record of another transaction. If so, instead of writing the rollback record, add the `has_overlapped_rollback` flag to that commit record. For example, when rolling a transaction whose `start_ts = 10` on a key, but there's another commit record whose `start_ts = 5` and `commit_ts = 10` that updates the key, if we write the rollback record directly, then the commit record will be overwritten. So instead, keep the commit record, but set the `has_overlapped_rollback` flag.
```

Comment:
I think we can make examples more concrete and state the order of commit and rollback, then it will be easier to understand.

For example, there are two transaction `T1 (start_ts = 10)` and `T2 (start_ts = 5)`. Both of them write key "k". First, `T2` commits key "k" with `commit_ts = 10`. After that, `T1` is going to rollback key "k". ...